Rounding

alexmemory opened the pull request.

alexmemory committed:
Adding fast method to calculate compat. of weighted logical rules.

alexmemory committed:
Added fast method for prob. a logical rule is satisfied.

Added a fast method for the probability a weighted logical rule is
satisfied, which is useful for rounding.  Also, changed the fast
calculation of expected compatibility, correcting the expectation of
the distance to satisfaction.

eriq-augustine commented on psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java:
```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+	 * were not grounded because they are trivially satisfied.
+	 */
+	public static double getExpectedWeightedLogicalCompatibility(WeightedGroundLogicalRule groundRule) {
+		/* Collects GroundAtom values. Atoms in GroundRule are
```
For comments in methods use line comments (//).

eriq-augustine commented on psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java:
```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+	 * from independently rounding each {@link RandomVariableAtom} to 1.0 or 0.0
+	 * with probability equal to its current truth value.
+	 *
+	 * WARNING: the result of this function is incorrect if the RandomVariableAtoms
```
Can we figure out a way around this?
I think this is too high a burden for the end user.
What is our use case? Do we want to round only one, or all the RVA?
If the later, then we can have a method that wraps this and resolves unweighted ground rules.

alexmemory replied:
Yes, we use this to round all RVA.  What does it mean to resolving unweighted ground rules?  My understanding is that the equivalence to MAXSAT does not hold in the case of unweighted rules.  This method (and others in the class) which are based on that equivalence have this restriction so that they are not used incorrectly.

eriq-augustine commented on psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java:
```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+	 * WARNING: the result of this function is incorrect if the RandomVariableAtoms
+	 * are subject to any {@link UnweightedGroundRule UnweightedGroundRule}.
+	 *
+	 * WARNING: This method does not account for WeightedGroundLogicalRules that
```
Why is this so?
Also, if this is true then we have to find a way around it.
A user will not always know up-front if they will have groundings in a rule (especially with lazy inference).

alexmemory replied:
I assumed this limitation still applied, but perhaps not.  At the very least, I wonder if it is wrong for non-lazy inference.

eriq-augustine commented on psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java:
```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+
+		// Sum of observed atoms
+		double k = 0;
+		for (double posObsValue : posObsValues)
```
Do this work in the loops above, then we don't have to store the constants in lists.

eriq-augustine commented on psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java:
```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+				posObsValues.add(atom.getValue());
+
+		// Sum of observed atoms
+		double k = 0;
```
More meaningful name, like observedSum or something.

eriq-augustine commented on psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java:
```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+		k = Math.min(1, k);
+
+		// P(sum of RV >= 1)
+		double notp = 1.0;
```
Better variable name.

eriq-augustine commented on psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java:
```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+
+		// P(sum of RV >= 1)
+		double notp = 1.0;
+		for (double posRvValue: posRvValues)
```
Compute in loops above so we don't have to store in lists.

eriq-augustine commented on psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java:
```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+			if (!(atom instanceof RandomVariableAtom) &&
+				(atom.getValue() > 0) &&
+				(atom.getValue() < 1))
+				throw new Exception("Soft observed values are not supported");
```
Throw a non-checked exception like an IllegalArgumentException.
Then you don't need to declare it as thrown.
Always avoid throwing checked exceptions when possible and appropriate.

eriq-augustine commented on psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java:
```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+			else
+				positiveValues[i] = posAtoms.get(i).getValue();
+		}
+		for (int i = 0; i < negativeValues.length; i++) {
```
I think this and the next set of loops can all be rolled into the first set of loops.
Then we don't have to allocate any lists/arrays.

eriq-augustine reviewed (changes requested):
I really need the theoretical writeup by Agelika (or you) before I can properly review this.

alexmemory committed:
Changed multi-line comments to line comments.

alexmemory committed:
Changed variable names to be more meaningful.

alexmemory committed:
Changed checked exception to unchecked exception.

Method getExpectedWeightedLogicalSatisfaction throws an unchecked
exception instead of a checked exception.

alexmemory committed:
Changed method to avoid allocating arrays of atoms.

Changed loops in getExpectedWeightedLogicalSatisfaction to avoid
allocating arrays of atoms.

alexmemory committed:
Changed method to avoid allocating arrays of atoms.

Changed loops in the getExpectedWeightedLogicalCompatibility to avoid
allocating lists of atoms.

Commit:
Rule names are no longer final

Commit:
Adding a fully ground rule causes a log entry, not an exception

eriq-augustine commented:
We take PSL PRs on the main repo now:
https://github.com/linqs/psl

I am very cautious about making rule names non-final.
What is the use case there?
With the online system coming in on the main repo (which needs to serialize rlules), I am concerned about rules changing names and not being properly recognized.

For the fully grounded rules, instead of throwing an exception or logging, I would like if we could just accept them.

alexmemory commented:
I forgot about this old PR and those commits shouldn't go back to the main repo.  So, please ignore the commits.  You had thoughtful follow-up questions, so I'll answer them anyway.

> I am very cautious about making rule names non-final.
> What is the use case there?

It is a work-around to add names to rules that are already created using org.linqs.psl.parser.ModelLoader.  I'm sure there is a better way to do that than this hack.

> With the online system coming in on the main repo (which needs to serialize rlules), I am concerned about rules changing names and not being properly recognized.

Yes, that sounds like a good reason to keep the names final.

> 
> For the fully grounded rules, instead of throwing an exception or logging, I would like if we could just accept them.

Agreed.  Last time this question came up years ago, it sounded like there was a lingering reason to assume rules were not ground at creation.  It would be nice if that were no longer true.

eriq-augustine commented:
For the pregrounded rules, I can't think of any deal-breaker reasons it can't be supported.
There would have to be some new infrastructure put in place, but nothing too crazy.
https://github.com/linqs/psl/issues/147
